Automattic/woocommerce-payments
# Fix invalid_request_error when level3 line_items contain a negative unit cost

Fixes [721](712-gh-Automattic/woocommerce-payments-server)

#### Changes proposed in this Pull Request
Fixes `invalid_request_error` when a payment_intent is created with level3 data. If the order contains a product that has a negative price, Stripe throws an error: `Invalid non-negative integer`.

The fix proposed would be to check if there is a `line_item` in the order with a negative price when the level3 data is calculated. When that is the case, we map the data in a different way:

1. Map the calculated `unit_cost` as `discount_amount`. We have to take the absolute value since Stripe only supports positive values
2. The `unit_cost` is set to 0

Using this approach the payment_intent will be created with the level3 data. The level3 data will include all the line_items  that have a negative price, but instead of having this amount in the `unit_cost`, it will be as `discount_amount`.









#### Testing instructions





1. Go to WP Admin > Products > Add New
2. Fill the product details, setting a negative amount in the Regular Price field, and Publish the product.
3. Create an order with some products, including the product created in previous step. Note that the total amount of the order needs to be greater than 0 to be able to check out.
4. Proceed to checkout and pay the order.
5. Check that the order is processed correctly.
6. In the Stripe Dashboard, check that the created payment contains the level3 data. Please check that the `line_items` are correct. The product with a negative price should show a `unit_cost` of 0 and the `discount_amount` should match the absolute value of the product price.

**Expected level3 data in Stripe's dashboard for the payment**

<img width="1190" alt="Screen Shot 2022-05-10 at 12 22 02" src="https://user-images.githubusercontent.com/1553182/167607890-eab8e263-3d69-4d24-9f28-5930a7ebba28.png">



-------------------


**Post merge**



- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : [Test Instructions](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-4.2.0#fix-invalid_request_error-for-payment_intents-caused-by-level3-data-unit_cost-is-negative-value-for-discount)
